From 8c4eb9b740a9f2ae57e048edbbc4aad1d62734f2 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 27 Feb 2024 15:04:35 +0800
Subject: [PATCH 094/104] mtk: hostapd: AP MLD: specify link id for unicast
 DEAUTH

When deauthenticating the STA, hostapd should specifies the setup link
of the target STA so that the TX status of the DEAUTH can be forwarded
to the correct link (BSS).

(The original gerrit somehow disappears, so I commit it again)
(https://gerrit.mediatek.inc/c/gateway/WiFi7/mac80211/hostapd/+/8715613)

CR-Id: WCNCR00289305
Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
Change-Id: I938f0fb80862074b95fc33b3c4f566f92ae21de1
---
 src/ap/ap_drv_ops.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 9357ce7b6..2c535f24a 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -867,7 +867,11 @@ int hostapd_drv_sta_deauth(struct hostapd_data *hapd,
 	if (hapd->conf->mld_ap) {
 		struct sta_info *sta = ap_get_sta(hapd, addr);
 
-		link_id = hapd->mld_link_id;
+		if (sta)
+			link_id = sta->mld_assoc_link_id;
+		else
+			link_id = hapd->mld_link_id;
+
 		if (ap_sta_is_mld(hapd, sta))
 			own_addr = hapd->mld->mld_addr;
 	}
-- 
2.39.2

